### PR TITLE
Write platform specific mute settings to backend

### PR DESCRIPTION
**Pull request by c3024**
### Details

Mute settings are currently saved only locally, and sometimes they get reset without any user action. This PR addresses the front-end implementation for saving platform-specific mute settings to the backend.

### Fixed Issues

$ https://github.com/Expensify/App/issues/49087
PROPOSAL: https://github.com/Expensify/App/issues/49087#issuecomment-2398351987

### Tests

1. Logon to ND on Web Chrome/Safari.
2. Click on "Settings" > "Preferences".
3. Toggle the "Mute all sounds from Expensify" switch to on.
4. Logout.
5. Logon again with the same account.
6. Click on "Settings" > "Preferences".
7. Verify that "Mute all sounds from Expensify" switch is **_on_**.
8. Logon with the same account on New Expensify Desktop app.
9. Click on "Settings" > "Preferences".
10. Verify that "Mute all sounds from Expensify" switch is **_off_**.
11. Follow steps 3 to 7 and verify the result on the Desktop app.
12. Logon to iOS New Expensify app.
13. Click on "Settings" > "Preferences".
14. Verify that "Mute all sounds from Expensify" switch is **_off_**.
15. Follow steps 3 to 7 and verify the result on iOS app.
16. Logon to Android New Expensify app.
17. Click on "Settings" > "Preferences".
18. Verify that "Mute all sounds from Expensify" switch is **_off_**.
19. Follow steps 3 to 7 and verify the result on Android app.

- [x] Verify that no errors appear in the JS console

### Offline tests

NA
### QA Steps

Same as Tests
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/02ec057f-6fb3-4d17-8d2c-12c8152560e1

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Not [tested](https://github.com/Expensify/App/pull/50493#issuecomment-2432497270) because there are still backend changes [required](https://github.com/Expensify/App/pull/50493#issuecomment-2432623982).

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/32a677f5-6586-4a61-8d8e-13dd56b8c0ed

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Not [tested](https://github.com/Expensify/App/pull/50493#issuecomment-2432497270) because there are still backend changes [required](https://github.com/Expensify/App/pull/50493#issuecomment-2432623982).

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/d33766f7-baf4-4d12-bcc3-a0a0f94d62fb

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/2908c56f-4ed6-4626-8b10-a65113dc449a

</details>

### CODE REVIEW

File: src/libs/actions/User.ts [hunk 1]
```diff
@@ -982,8 +986,31 @@ function clearUserErrorMessage() {
     Onyx.merge(ONYXKEYS.USER, {error: ''});
 }
 
-function setMuteAllSounds(isMutedAllSounds: boolean) {
-    Onyx.merge(ONYXKEYS.USER, {isMutedAllSounds});
+function togglePlatformMute(platform: Platform, mutedPlatforms: Platform[]) {
+    const isPlatformMuted = mutedPlatforms?.includes(platform);
+    const newMutedPlatforms = isPlatformMuted ? mutedPlatforms.filter((mutedPlatform) => mutedPlatform !== platform) : [...mutedPlatforms, platform];
```
Comment (tgolen) [inline src/libs/actions/User.ts:991]: Ah, sorry. I guess this was something I didn't explicitly say. It's actually an object like this:

```
{
    'web': true,
    'desktop': true,
}
```

and if a platform is muted, it is removed from the object entirely.

**Review by tgolen — CHANGES_REQUESTED** [review]

File: src/libs/actions/User.ts (hunk 1 above)
Comment (c3024) [reply src/libs/actions/User.ts]: Done!

File: src/libs/actions/User.ts [hunk 2]
```diff
@@ -982,8 +986,32 @@ function clearUserErrorMessage() {
...
-function setMuteAllSounds(isMutedAllSounds: boolean) {
-    Onyx.merge(ONYXKEYS.USER, {isMutedAllSounds});
+function togglePlatformMute(platform: Platform, mutedPlatforms: Partial<Record<Platform, true>>) {
+    const newMutedPlatforms = mutedPlatforms?.[platform]
+        ? {...mutedPlatforms, [platform]: undefined} // Remove platform if it's muted
+        : {...mutedPlatforms, [platform]: true}; // Add platform if it's not muted
+
+    const optimisticData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.SET,
```
Comment (c3024) [inline src/libs/actions/User.ts:996]: Is it better to use `MERGE` here without passing `mutedPlatforms` to this function `togglePlatformMute`?

Comment (tgolen) [reply src/libs/actions/User.ts:996]: For NVPs, the server uses SET, so it's probably best to use SET here as well. That will keep things consistent.

**Comment by tgolen** [comment]
Hi! The API has been deployed to staging, so you should be able to finish this up now.

**Comment by c3024** [comment]
When we mute some platforms and then unmute all of them, backend seems to send an empty array and Onyx complains.

Screenshot 2024-10-16 at 8 22 20 AM

**Comment by c3024** [comment]
`OpenApp` uses `merge` for this NVP. There’s no issue with this, per se, but should we use `set` here instead?

Screenshot 2024-10-16 at 8 28 28 AM

**Comment by tgolen** [comment]
> When we mute some platforms and then unmute all of them, backend seems to send an empty array in Pusher to other devices (here it's and iOS emulator) and Onyx complains.

Ah, this is because of PHP's JSON encoding which converts the empty object to an empty array. Very troublesome. I think it might be better to set the value to `null`. I can make that BE change, but I wouldn't let it block this PR for now.

> OpenApp uses merge for this NVP. There’s no issue with this, per se, but should we use set here instead?

It looks like it's the same for all the NVPs, so I think that's OK for now. I agree it's not the most consistent, but I don't really want to change the behavior of all the NVPs in `OpenApp` unless there is a clear problem.

**Comment by tgolen** [comment]
OK, I merged a fix for the empty array. It should deploy to staging tomorrow. Here, you can see me making the request in PostMan and then NewDot receiving the Onyx update. I hope that will work for you!

image

**Comment by tgolen** [comment]
Hm, no... They shouldn't be. I can look into that. Let's not let that block this PR though.

**Comment by c3024** [comment]
@tgolen 

>Let's not let that block this PR though.

🤔 But, without fixing that we cannot reliably test this PR.

1. Log on to mWeb.
2. Frontend looks for an `mobileWeb` field in the nvp and it does not exist so the mute settings value is `false`.
3. Switch it on. Optimistically it updates. But, backend immediately returns a response again with the nvp without `mobileWeb` field. So, it switches back off. As it exists now, we can never switch the mute setting to on on `mWeb` unless I also merge `web` and `mWeb` on frontend.

**Comment by tgolen** [comment]
Yeah, I understand it would be a bug on mobile web, but mobile web was also the least important (someone even proposed not supporting mobile web at all). It's up to you though. I can probably get a fix written for it today or tomorrow, and that would be merged and deployed by Monday if you want to wait.

**Comment by c3024** [comment]
Thanks. Going ahead!

**Comment by tgolen** [comment]
@c3024 I looked into that bug and found some interesting things. I wasn't able to reproduce it using PostMan. These are the four requests I made:

- Toggle desktop
- Toggle web
- Toggle mobile web
- Toggle mobile web

and here were the OnyxUpdates sent to the client via pusher (which look correct)

image

When I looked at the logs for your request ID `8d3e4cc1b97885fd-BOM` it shows that the `platform` param actually came through as `web`:

> `Processing 'TogglePlatformMute' for 'expensify.com' from '122.179.38.14' ~~ browserGUID: '670df802a8aeb' initialReferer: '<REDACTED>' command: 'TogglePlatformMute' authToken: '<REDACTED>' platform: 'web' pusherSocketID: '788900.884744' referer: 'ecash' `


### Conclusion
It looks like this is a frontend issue, so you might want to check the actual network requests to see what parameter is sent to the server.

**Comment by c3024** [comment]
@tgolen 

Apparently, we enhance the request parameters and add platform. So, `platform` param of the request is overwritten here because `getPlatform()` treats `mWeb` also as `web`.

https://github.com/Expensify/App/blob/866ce4b3957111d2c3ea5a7dccc3c86b90ed4e7c/src/libs/Network/enhanceParameters.ts#L44

We need to rename the `platform` param for this command `TogglePlatformMute`. Let me know what name will be used in the backend.

**Comment by jjcoffee** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by c3024** [comment]
Sounds good!

**Comment by tgolen** [comment]
Great, I've submitted a PR to the server for that so it should be live and on staging by tomorrow I expect.

**Comment by jjcoffee** [comment]
@c3024 Were you able to retest with the staging server?

**Comment by c3024** [comment]
@jjcoffee 

Backend PR has not been deployed to Staging yet.


https://github.com/user-attachments/assets/41d849b9-e156-49ce-bdd3-65cd8f33b54d

**Comment by tgolen** [comment]
Yeah, it still hasn't been deployed yet. I'm hopeful that will happen today. I'll try to ping here as soon as I see it deployed.

**Review by jjcoffee — APPROVED** [review]
Tests nicely using the staging server!

**Comment by tgolen** [comment]
Cool, thanks! I see the backend PR was deployed about 17 hours ago.

**Review by tgolen — APPROVED** [review]

**Pull request merged**